src/listener/Heading.php

```diff
@@ -45,21 +45,44 @@ public function render(Lexer $lexer)
                 // prevent html injection in case the attribute is user input
                 throw new Exception('An unknown heading level "'.$pick->heading.'" has been detected.');
             }
-            
-            // get all
-            $prev = $pick->line->previous(function (Line $line) {
-                if (!$line->isInline()) {
-                    return true;
-                }
-            });
-
-            // if there is no previous element, we take the same line element.
-            if (!$prev) {
-                $prev = $pick->line;
-            }
-
-            $pick->line->output = '<h'.$pick->heading.'>'.$prev->getInput() . $pick->line->renderPrepend() . '</h'.$pick->heading.'>';
-            $prev->setDone();
-        }
+	
+	        $first = $this->getFirstLine($pick);
+	
+	        // while from first to the pick line and store content in buffer
+            $buffer = null;
+            $first->while(function (&$index, Line $line) use (&$buffer, $pick) {
+				$index++;
+				$buffer.= $line->getInput();
+				$line->setDone();
+				if ($index == $pick->line->getIndex()) {
+					return false;
+				}
+			});
+	
+			$pick->line->output = '<h'.$pick->heading.'>'.$buffer . '</h'.$pick->heading.'>';
+			$pick->line->setDone();
+		}
     }
+
+	private function getFirstLine($pick) {
```

Comment:
Even its a private method, we need php docs including type hinting of $pick object and since tags. Also please take a look at coding styles as mentioned.

Maybe we could even provide such methods as part of base Listener object, cause we should avoid using the same code twice (DRY concept).